core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -51,8 +185,744 @@ trait DataSkippingReader
   private lazy val withStatsCache =
     cacheDS(withStatsInternal0, s"Delta Table State with Stats #$version - $redactedPath")
 
-  // TODO(scott): proper implementation
+  protected def withStatsInternal: DataFrame = withStatsCache.getDS
+
+  /** All files with the statistics column dropped completely. */
+  def withNoStats: DataFrame = allFiles.drop("stats")
+
+  /**
+   * Returns a parsed and cached representation of files with statistics.
+   *
+   *
+   * @return cached [[DataFrame]]
+   */
+  final def withStats: DataFrame = {
+    withStatsInternal
+  }
+
+
+  // Helper method for expression types that represent an IN-list of literal values.
+  //
+  //
+  // For excessively long IN-lists, we just test whether the file's min/max range overlaps the range
+  // spanned by the list's smallest and largest elements.
```

Comment:
We should set a threshold, when `in.size` is less than the threshold, we should convert `In` to a series of `Equals`, and then use the filtering rules of `Equals`

For instance:


```
col_name in (v1, v2) => col_name == v1 or col_name == v2
```

> suppose `v1 < v2`

the file skipping predicate expression before:

```
stats.minValues.col_name <= v1 and stats.maxValues.col_name >= v2
```

the file skipping predicate expression after:
```
(stats.minValues.col_name <= v1 and stats.maxValues.col_name >= v1)
  or (stats.minValues.col_name <= v2 and stats.maxValues.col_name >= v2)
```

the second one has a better file filtration efficiency

Reply:
Hi @cfmcgrady, that's a great suggestion.

I think that this PR is big enough already ... once it is merged, would you like to implement your proposed solution? 

Or perhaps, at the very least, create an issue documenting with more detail what you'd like to implement + why it is mathematically correct (these changes can get rather complicated).

Reply:
@cfmcgrady want to make an issue for this improvement?